core/crypto/kem/hybrid/hybrid.go

```diff
@@ -0,0 +1,334 @@
+// so that a non-uniform seed (such as a shared secret generated by a hybrid
+// KEM where one of the KEMs is weak) doesn't impact just one of the KEMs.
+//
+// Of our XOF (SHAKE256), we desire two security properties:
+//
+//  1. The internal state of the XOF should be big enough so that we
+//     do not loose entropy.
+//  2. From one of the new seeds, we shouldn't be able to derive
+//     the other or the original seed.
+//
+// SHAKE256, and all siblings in the SHA3 family, have a 200B internal
+// state, so (1) is fine if our seeds are less than 200B.
+// If SHAKE256 is computationally indistinguishable from a random
+// sponge, then it affords us 256b security against (2) by the
+// flat sponge claim [https://keccak.team/files/SpongeFunctions.pdf].
+// None of the implemented schemes claim more than 256b security
+// and so SHAKE256 will do fine.
+package hybrid
+
+import (
+	"errors"
+
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/cloudflare/circl/kem"
+)
+
+const (
+	// SeedSize is the number of bytes needed to seed deterministic methods below.
+	SeedSize = 32
+)
+
+var (
+	ErrUninitialized = errors.New("public or private key not initialized")
+)
+
+// Public key of a hybrid KEM.
+type PublicKey struct {
+	scheme *Scheme
+	first  kem.PublicKey
+	second kem.PublicKey
+}
+
+// Private key of a hybrid KEM.
+type PrivateKey struct {
+	scheme *Scheme
+	first  kem.PrivateKey
+	second kem.PrivateKey
+}
+
+// Scheme for a hybrid KEM.
+type Scheme struct {
+	name   string
+	first  kem.Scheme
+	second kem.Scheme
+}
+
+// New creates a new hybrid KEM given the first and second KEMs.
+func New(name string, first kem.Scheme, second kem.Scheme) *Scheme {
+	return &Scheme{
+		name:   name,
+		first:  first,
+		second: second,
+	}
+}
+
+func (sch *Scheme) Name() string { return sch.name }
+func (sch *Scheme) PublicKeySize() int {
+	return sch.first.PublicKeySize() + sch.second.PublicKeySize()
+}
+
+func (sch *Scheme) PrivateKeySize() int {
+	return sch.first.PrivateKeySize() + sch.second.PrivateKeySize()
+}
+
+func (sch *Scheme) SeedSize() int {
+	return SeedSize
+}
+
+func (sch *Scheme) SharedKeySize() int {
+	return sch.first.SharedKeySize() + sch.second.SharedKeySize()
+}
+
+func (sch *Scheme) CiphertextSize() int {
+	return sch.first.CiphertextSize() + sch.second.CiphertextSize()
+}
+
+func (sch *Scheme) EncapsulationSeedSize() int {
+	return SeedSize
+}
+
+func (sk *PrivateKey) Scheme() kem.Scheme { return sk.scheme }
+func (pk *PublicKey) Scheme() kem.Scheme  { return pk.scheme }
+
+func (sk *PrivateKey) MarshalBinary() ([]byte, error) {
+	if sk.first == nil || sk.second == nil {
+		return nil, ErrUninitialized
+	}
+	first, err := sk.first.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	second, err := sk.second.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	return append(first, second...), nil
+}
+
+func (sk *PrivateKey) Equal(other kem.PrivateKey) bool {
```

Comment:
it looks like they are handling pointer nil semantics of equality here. i suppose they had some uses in mind like:

```
var myKey PrivateKey = nil

// marklar marklar...

if myKey == nil {
        // marklar
}

// marklar 

if myKey.Equal(otherKey) {
        // marklar
}
```

Reply:
lol maybe they added that stuff in to prevent a bunch of panics... because golang runtime will panic if you try to compare a byte slice if it is set to nil.

Reply:
so... yeah we can remove that crufty code... but there might be a hidden state machine somewhere that later crashes with a panic inside the Equal method... with a nil pointer dereference error traceback ;-p

Reply:
but in defense of your suggestion that we may not need it... we don't have such checks anywhere else in our code and we've been fine.